### ISSUE TITLE
Merge generating too many parquet files 

### ISSUE THREAD

**Issue by kkr78** [delta-io/delta#239]
The merge is generating too many parquet files in the output directory.  is there any way to control?

**Comment by sanjiv1980** [comment]
Hey, I have also experienced the same and about to ask this question . There is compaction API which do the merging but again , it's all together different .

**Comment by tdas** [comment]
the only control we have for now is to control the shuffle partitions, that is, SQL configuration `spark.sql.shuffle.partitions`. Other than that, the main solution is to [compact the file](https://docs.delta.io/0.4.0/best-practices.html#compact-files). This will easier with [0.5.0](https://github.com/delta-io/delta/milestone/5) because we will be [enabling a way](https://github.com/delta-io/delta/issues/146) to specify that a compaction job does not chanage data (i.e., option `dataChange = false`) such that the compaction job does not conflict with other concurrent jobs.

**Comment by sanjiv1980** [comment]
@tdas My compaction job got failed , Same configuration I have used to formed `Delta Lake` with more than 300GB of data , thereafter I performed some merge operation, which opened so many small files . When I ran the compaction job, It was keep on going for more than 6-hours. Here is my code :

```
spark.read
  .format("delta")
  .load(path)
  .where(partition)
  .repartition(numFilesPerPartition)
  .write
  .format("delta")
  .mode("overwrite")
  .option("replaceWhere", partition)
  .save(path)```

Please let me know If I am doing anything wrong . One-thing I would like to mentioned , My reading and writing directory are same .

**Comment by marmbrus** [comment]
> One-thing I would like to mention ,Like my reading and writing directory is same
This is okay, due to delta's snapshot isolation.

What is the error you got?

**Comment by sanjiv1980** [comment]
There is no such error , but my all executors are getting killed after running sometimes and job is still running since more than 7 hours . I saw data shuffled around 2.4TB (My delta lake volume is only 300GB).

**Comment by marmbrus** [comment]
It is not unusual for the shuffle size to be significantly larger than the data stored as parquet.  Compression and columnar tricks like RLE can dramatically reduce the data size, where as the shuffle is row-wise tungsten encoded data (designed for speed of access rather than minimal data size).

It is hard to say what is going wrong here without seeing the error logs, but my guess would be 
this is not delta specific and you are just running out of disk space. Some modes of running Spark use [autoscaling instance storage to avoid this type of failure](https://databricks.com/blog/2017/12/01/transparent-autoscaling-of-instance-storage.html)

**Comment by kkr78** [comment]
> spark.sql.shuffle.partitions

This worked fine. I see very few files now.

**Comment by sanjiv1980** [comment]
@marmbrus I am getting below error :
```
19/11/14 08:06:06 INFO UnsafeExternalSorter: Thread 67 spilling sort data of 3.0 GB to disk (0  time so far)
19/11/14 08:06:14 INFO UnsafeExternalSorter: Thread 57 spilling sort data of 3.0 GB to disk (0  time so far)
19/11/14 08:06:24 INFO UnsafeExternalSorter: Thread 66 spilling sort data of 3.0 GB to disk (0  time so far)
19/11/14 08:06:26 INFO UnsafeExternalSorter: Thread 55 spilling sort data of 3.0 GB to disk (0  time so far)
19/11/14 08:08:00 INFO UnsafeExternalSorter: Thread 66 spilling sort data of 3.0 GB to disk (1  time so far)
19/11/14 08:08:08 INFO UnsafeExternalSorter: Thread 67 spilling sort data of 3.0 GB to disk (1  time so far)
19/11/14 08:08:13 INFO UnsafeExternalSorter: Thread 55 spilling sort data of 3.0 GB to disk (1  time so far)
19/11/14 08:08:17 INFO UnsafeExternalSorter: Thread 57 spilling sort data of 3.0 GB to disk (1  time so far)
19/11/14 08:09:59 INFO UnsafeExternalSorter: Thread 66 spilling sort data of 3.0 GB to disk (2  times so far)
19/11/14 08:10:09 INFO UnsafeExternalSorter: Thread 57 spilling sort data of 3.0 GB to disk (2  times so far)
19/11/14 08:10:14 INFO UnsafeExternalSorter: Thread 55 spilling sort data of 3.0 GB to disk (2  times so far)
19/11/14 08:10:49 INFO UnsafeExternalSorter: Thread 67 spilling sort data of 3.0 GB to disk (2  times so far)
19/11/14 08:11:51 INFO UnsafeExternalSorter: Thread 66 spilling sort data of 3.0 GB to disk (3  times so far)
19/11/14 08:12:22 INFO UnsafeExternalSorter: Thread 57 spilling sort data of 3.0 GB to disk (3  times so far)
19/11/14 08:12:29 INFO UnsafeExternalSorter: Thread 55 spilling sort data of 3.0 GB to disk (3  times so far)
19/11/14 08:13:20 INFO UnsafeExternalSorter: Thread 67 spilling sort data of 3.0 GB to disk (3  times so far)
19/11/14 08:13:55 INFO UnsafeExternalSorter: Thread 66 spilling sort data of 3.0 GB to disk (4  times so far)
19/11/14 08:14:07 INFO CoarseGrainedExecutorBackend: Driver commanded a shutdown
19/11/14 08:14:07 ERROR TransportResponseHandler: Still have 16 requests outstanding when connection from ip-172-31-40-231.ec2.internal/172.31.40.231:7337 is closed
19/11/14 08:14:07 ERROR TransportResponseHandler: Still have 3 requests outstanding when connection from ip-172-31-43-200.ec2.internal/172.31.43.200:7337 is closed
19/11/14 08:14:07 INFO RetryingBlockFetcher: Retrying fetch (1/3) for 3 outstanding blocks after 5000 ms
19/11/14 08:14:07 INFO RetryingBlockFetcher: Retrying fetch (1/3) for 5 outstanding blocks after 5000 ms
19/11/14 08:14:07 INFO RetryingBlockFetcher: Retrying fetch (1/3) for 6 outstanding blocks after 5000 ms
19/11/14 08:14:07 INFO RetryingBlockFetcher: Retrying fetch (1/3) for 2 outstanding blocks after 5000 ms
19/11/14 08:14:07 ERROR OneForOneBlockFetcher: Failed while starting block fetches
java.io.IOException: Connection from ip-172-31-40-231.ec2.internal/172.31.40.231:7337 closed
	at org.apache.spark.network.client.TransportResponseHandler.channelInactive(TransportResponseHandler.java:146)
	at org.apache.spark.network.server.TransportChannelHandler.channelInactive(TransportChannelHandler.java:108)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
	at io.netty.handler.timeout.IdleStateHandler.channelInactive(IdleStateHandler.java:277)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
	at org.apache.spark.network.util.TransportFrameDecoder.channelInactive(TransportFrameDecoder.java:182)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1354)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
	at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:917)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:822)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Thread.java:748)
19/11/14 08:14:07 INFO RetryingBlockFetcher: Retrying fetch (1/3) for 5 outstanding blocks after 5000 ms
19/11/14 08:14:07 ERROR OneForOneBlockFetcher: Failed while starting block fetches```

**Comment by sanjiv1980** [comment]
One more thing I observed like after running the compaction API , file size is getting increased . My directory file size was 100 GB , after compaction it has increased by 5-times , now I am seeing it 500GB.

**Comment by mukulmurthy** [comment]
I'd expect directory size to roughly double, because you're reading all the data and writing it back out. VACUUM should clean up the files that are no longer part of the table. What was your retention period you ran VACUUM with? By default, it keeps data files that were removed from the table within the last 7 days, so if you ran VACUUM right after your compaction it wouldn't clean up the old files.

**Comment by sanjiv1980** [comment]
@mukulmurthy , But my file size is 5 times bigger than that, It was 100GB , now its 500GB. My retention period is less than 10 sec . I checked my deleted files are getting removed from directory. Could you please explain me why the size has increased .? (and that more than 5 times bigger).

**Comment by mukulmurthy** [comment]
I'd try running VACUUM with retention of 0. If it's not that, you'd have to look and see the size of the actual data files and where that extra data is coming from. I'm not sure I'd be able to assist further with that; it's very hard to debug these kinds of issues without looking at the actual storage.

**Comment by sanjiv1980** [comment]
@mukulmurthy Actually, I was trying to fixed the file size , by calculating the repartition 
(total-data-volume/size-that-I-need), but unable to do that because `total-data-volume` are getting increased , so unable to stick with my size what I expected.

**Comment by tdas** [comment]
The original issue of too many files being generated during merge has been addressed in #349, to be released in 0.6.0. Closing this issue.

**Comment by adarsh112** [comment]
You can set repartitionBeforeWrite=True if you are using older deltalake version. In the newer versions it is enabled by default
